torchrl/modules/distributions/continuous.py

```diff
@@ -392,22 +426,71 @@ def update(self, loc: torch.Tensor, scale: torch.Tensor) -> None:
 
         if (
             hasattr(self, "base_dist")
-            and (self.base_dist.base_dist.loc.shape == self.loc.shape)
-            and (self.base_dist.base_dist.scale.shape == self.scale.shape)
+            and (self.root_dist.loc.shape == self.loc.shape)
+            and (self.root_dist.scale.shape == self.scale.shape)
         ):
-            self.base_dist.base_dist.loc = self.loc
-            self.base_dist.base_dist.scale = self.scale
+            self.root_dist.loc = self.loc
+            self.root_dist.scale = self.scale
         else:
-            base = D.Independent(D.Normal(self.loc, self.scale), self._event_dims)
-            super().__init__(base, self._t)
+            if self._event_dims > 0:
+                base = D.Independent(D.Normal(self.loc, self.scale), self._event_dims)
+                super().__init__(base, self._t)
+            else:
+                base = D.Normal(self.loc, self.scale)
+                super().__init__(base, self._t)
+
+    @property
+    def root_dist(self):
+        bd = self
+        while hasattr(bd, "base_dist"):
+            bd = bd.base_dist
+        return bd
 
     @property
     def mode(self):
-        m = self.base_dist.base_dist.mean
+        def newton_raphson_step(x, alpha=1):
+            # find where gradient is approx 0
+            f = lambda x: self.log_prob(x).mean()
+
+            def grad(x):
+                return torch.func.grad(f)(x)
+
+            def hess(x):
+                return torch.func.grad_and_value(grad)(x)
+
+            h, g = hess(x)
+            step = g / h
+            return (alpha * step).detach()
+
+        # Get starting point
+        m = self.root_dist.mean
         for t in self.transforms:
             m = t(m)
+
+        # m is our start point
+        for alpha in torch.arange(0.9, 0, -0.05):
+            m = m.clone().requires_grad_()
+            newton_raphson_step_vmap = lambda x: newton_raphson_step(x, alpha=alpha)
+            for i in range(m.ndim):
+                newton_raphson_step_vmap = torch.vmap(newton_raphson_step_vmap)
+            step = newton_raphson_step_vmap(m)
+            m = m.detach()
+            # Correct the step
+            step = (m + step).clamp(self.low, self.high) - m
+            # make the step
+            m = m + step
+            if abs(step).max() < 1e-3:
+                break
+        else:
+            torchrl_logger.info(
+                f"Failed to find the mode of the TanhNormal distribution. Max step={abs(step).max()}"
+            )
         return m
 
+    @property
+    def mean(self):
+        return self.rsample((1000,)).mean(0)
```

Comment:
This is a bit arbitrary no? Isn't it better to not provide this feature and tell users that they can compute their empirical estimate themselves?